add grouped-and-auto-accessors (#255)
estree/estree: experimental/grouped-and-auto-accessors.md

# [Grouped Accessors and Auto-Accessors][proposal-grouped-and-auto-accessors]

## ClassBody

```js
extend interface ClassBody {
    body: [ MethodDefinition | PropertyDefinition | StaticBlock | ClassAccessor ];
}
```

## ObjectExpression

```js
extend interface ObjectExpression {
    properties: [ Property | ObjectAccessor ];
}
```

## ClassAccessor

```js
interface ClassAccessor <: PropertyDefinition {
    type: "ClassAccessor";
    get: GetAccessorStub | GetAccessorMethod | null;

>     set: SetAccessorStub | SetAccessorMethod | null;

Reviewer: `ClassAccessor` should also have a `value` field which is similar to a class field's value/initializer:

```js
accessor foo { get; set; } = 123;
```

Reviewer: Ah, I see, it inherits these properties. I think this makes sense.

}
```

> If `set.private` is `true`, `key` must be an `Identifier` and `computed` must be `false`.

Reviewer 2: Maybe this is a bit clearer?

```suggestion
If `set.private` is `true`, `ClassAccessor#key` must be an `Identifier` and `ClassAccessor#computed` must be `false`.
```

## ObjectAccessor

```js
interface ObjectAccessor <: Property {
    type: "ObjectAccessor";
    get: GetAccessorStub | GetAccessorMethod | null;
    set: SetAccessorStub | SetAccessorMethod | null;
}
```

> ### GetAccessorStub

Reviewer 2: Why use “stub” instead of using the proposal language, such as GetAutoAccessor?

Author: The "stub" comes from the [proposed syntax section](https://github.com/tc39/proposal-grouped-and-auto-accessors#proposed-syntax).

According to the spec, an auto-accessor is a simplified version of grouped accessor, e.g. `accessor x = { get; set; }`, where the accessor stub refers to the `set` and `get` shorthand.

Reviewer 2: Ah ok.

```js
interface GetAccessorStub <: Node {
    type: "GetAccessorStub";
}
```

> A get accessor stub is a simplified version of grouped accessor. E.g. `class C { accessor c { get; } }`.

Author: ```suggestion
A get accessor stub is used in auto-accessor. E.g. `class C { accessor c { get; } }`.
```

### GetAccessorMethod

```js
interface GetAccessorMethod <: Node {
    type: "GetAccessorMethod";

>     params: [];

Author: The `params` must be an empty array. It is to provide compatibility for general function AST visitors: A function always has `params` property.

    body: FunctionBody;
}
```

### SetAccessorStub

```js
interface SetAccessorStub <: Node {
    type: "SetAccessorStub";
    private: boolean;
}
```

> A set accessor stub is a simplified version of grouped accessor. E.g. `class C { accessor c { set; } }`.

Author: ```suggestion
A set accessor stub is used in auto-accessor. E.g. `class C { accessor c { set; } }`.
```

# [Grouped Accessors and Auto-Accessors][proposal-grouped-and-auto-accessors]

## ClassBody

```js
extend interface ClassBody {
    body: [ MethodDefinition | PropertyDefinition | StaticBlock | ClassAccessor ];
}
```

## ObjectExpression

```js
extend interface ObjectExpression {
    properties: [ Property | ObjectAccessor ];
}
```

## ClassAccessor

```js
interface ClassAccessor <: PropertyDefinition {
    type: "ClassAccessor";
    get: GetAccessorStub | GetAccessorMethod | null;
    set: SetAccessorStub | SetAccessorMethod | null;
}
```

If `set.private` is `true`, `ClassAccessor#key` must be an `Identifier` and `ClassAccessor#computed` must be `false`.

## ObjectAccessor

```js
interface ObjectAccessor <: Node {
    type: "ObjectAccessor";
    key: Literal | Identifier;
    computed: boolean;
    get: GetAccessorStub | GetAccessorMethod | null;
    set: SetAccessorStub | SetAccessorMethod | null;

> }

Author: In the latest commit I removed the inheritance from `Property`, because it will inherit `value: Expression` and `kind: "init" | "set" | "get"`. These properties are not application to the grouped accessor.

Reviewer 2: Makes sense.
